statistics/handle/handle.go

```diff
@@ -647,64 +647,128 @@ func (h *Handle) LoadNeededHistograms() (err error) {
 	}()
 	loadFMSketch := config.GetGlobalConfig().Performance.EnableLoadFMSketch
 
-	for _, col := range cols {
-		oldCache := h.statsCache.Load().(statsCache)
-		tbl, ok := oldCache.Get(col.TableID)
-		if !ok {
-			continue
-		}
-		c, ok := tbl.Columns[col.ColumnID]
-		if !ok || !c.IsLoadNeeded() {
-			statistics.HistogramNeededColumns.Delete(col)
-			continue
+	for _, item := range items {
+		if !item.IsIndex {
+			err = h.loadNeededColumnHistograms(reader, item, loadFMSketch)
+		} else {
+			err = h.loadNeededIndexHistograms(reader, item, loadFMSketch)
 		}
-		hg, err := h.histogramFromStorage(reader, col.TableID, c.ID, &c.Info.FieldType, c.Histogram.NDV, 0, c.LastUpdateVersion, c.NullCount, c.TotColSize, c.Correlation)
 		if err != nil {
-			return errors.Trace(err)
+			return err
 		}
-		cms, topN, err := h.cmSketchAndTopNFromStorage(reader, col.TableID, 0, col.ColumnID)
+	}
+	return nil
+}
+
+func (h *Handle) loadNeededColumnHistograms(reader *statsReader, col model.TableItemID, loadFMSketch bool) (err error) {
+	oldCache := h.statsCache.Load().(statsCache)
+	tbl, ok := oldCache.Get(col.TableID)
+	if !ok {
+		return nil
+	}
+	c, ok := tbl.Columns[col.ID]
+	if !ok || !c.IsLoadNeeded() {
+		statistics.HistogramNeededItems.Delete(col)
+		return nil
+	}
+	hg, err := h.histogramFromStorage(reader, col.TableID, c.ID, &c.Info.FieldType, c.Histogram.NDV, 0, c.LastUpdateVersion, c.NullCount, c.TotColSize, c.Correlation)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	cms, topN, err := h.cmSketchAndTopNFromStorage(reader, col.TableID, 0, col.ID)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	var fms *statistics.FMSketch
+	if loadFMSketch {
+		fms, err = h.fmSketchFromStorage(reader, col.TableID, 0, col.ID)
 		if err != nil {
 			return errors.Trace(err)
 		}
-		var fms *statistics.FMSketch
-		if loadFMSketch {
-			fms, err = h.fmSketchFromStorage(reader, col.TableID, 0, col.ColumnID)
-			if err != nil {
-				return errors.Trace(err)
-			}
-		}
-		rows, _, err := reader.read("select stats_ver from mysql.stats_histograms where is_index = 0 and table_id = %? and hist_id = %?", col.TableID, col.ColumnID)
+	}
+	rows, _, err := reader.read("select stats_ver from mysql.stats_histograms where is_index = 0 and table_id = %? and hist_id = %?", col.TableID, col.ID)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if len(rows) == 0 {
+		logutil.BgLogger().Error("fail to get stats version for this histogram", zap.Int64("table_id", col.TableID), zap.Int64("hist_id", col.ID))
```

Comment:
Maybe we should return error here otherwise panic happens when accessing `rows[0]`.